File: tools/fontgen/export_glyphs.py
"""Export a fontgen GlyphSet to JSON for node-side consumers
(guess-letter.mjs, blind-read.mjs).

Both y-phases are exported (integer and half-px baselines — MuPDF can produce
either; the known corpus uses integer only). Rasters are the raw uint8 gray
windows (single glyph on white), base64, row-major, with (dx, dy) offsets
relative to the integer pen / baseline and the exact dyadic freetype advance.
Keys are "phx_phy" (e.g. "0.25_0.5").

    python tools/fontgen/export_glyphs.py [assets/fonts/times_16.npz] [out.json]
"""
import base64, json, sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from fontgen import GlyphSet

if len(sys.argv) == 2:
    sys.exit(f"refusing to write {sys.argv[1]} to the default glyphs_times16.json - "
             f"pass an explicit out path, e.g.\n"
             f"    python tools/fontgen/export_glyphs.py {sys.argv[1]} "
             f"assets/glyphs/glyphs_<name>.json")

REPO = Path(__file__).resolve().parents[2]
npz = Path(sys.argv[1]) if len(sys.argv) > 1 else REPO / "assets" / "fonts" / "times_16.npz"
out = Path(sys.argv[2]) if len(sys.argv) > 2 else REPO / "assets" / "glyphs" / "glyphs_times16.json"

gs = GlyphSet.load(npz)
chars = {}
for c in gs.meta["chars"]:
    ph = {}
    for phx in gs.meta["phases_x"]:
        for phy in gs.meta["phases_y"]:
            arr, dx, dy = gs.get(c, phx, phy)
            key = str(phx) if phy == 0.0 else f"{phx}_{phy}"   # legacy key for phy=0
            ph[key] = {"w": int(arr.shape[1]) if arr.size else 0,
                       "h": int(arr.shape[0]) if arr.size else 0,
                       "dx": int(dx), "dy": int(dy),
                       "b64": base64.b64encode(arr.tobytes()).decode() if arr.size else ""}
    chars[c] = {"adv": float(gs.advances[c]), "ph": ph}
out.write_text(json.dumps({"font": str(npz.name), "size_px": gs.meta["size_px"],
                           "linear": "linear-remap" in gs.meta.get("pipeline", ""),
                           "phases_x": gs.meta["phases_x"],
                           "phases_y": gs.meta["phases_y"], "chars": chars}))
print(f"{out}: {len(chars)} chars x {len(gs.meta['phases_x'])}x{len(gs.meta['phases_y'])} phases, "
      f"{out.stat().st_size // 1024} KB")

File: tools/export-glyphs.mjs
// ---------------------------------------------------------------------------
// export-glyphs.mjs — build assets/glyphs/glyphs.bin, THE glyph dictionary
// (every fontgen set in one binary file), from the committed .npz rasters
// (assets/fonts/, zero corpus pixels). Layout: tools/glyph-bundle.mjs.
// Zero dependencies (own zip/npy parser).
//
//   node tools/export-glyphs.mjs            # (re)build glyphs.bin
//   node tools/export-glyphs.mjs --check    # rebuild in memory and byte-
//         compare against the committed bundle — proves it is a pure
//         derivation of the committed rasters (npm run glyphs-check)
//
// The set manifest lives in tools/glyph-registry.mjs (a new font = new line there).
// Both y-phases are exported (integer and half-px baselines); rasters are
// the raw uint8 MuPDF gray windows plus the true rasterizer alpha derived
// through the set's compositor law (see COV), with (dx, dy) offsets
// relative to the integer pen / baseline and the exact dyadic freetype
// advance. The per-set JSON era (glyphs_*.json) ended 2026-07-16; the one
// parked _OFF experiment keeps its JSON as provenance.
// ---------------------------------------------------------------------------
import { readFileSync, writeFileSync } from 'node:fs';
import { join, resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { buildSetFromNpz } from './glyph-bundle.mjs';
import { SETS } from './glyph-registry.mjs';

const REPO = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const FONTS = join(REPO, 'assets', 'fonts');
const BUNDLE = join(REPO, 'assets', 'glyphs', 'glyphs.bin');

// The name → npz manifest is SETS in tools/glyph-registry.mjs — ONE registry
// for sets, family pools, and rosters (a new font = one line there).


// zip/npy parsing, the alpha law, and payload building are SHARED with the
// readers' direct-.npz loading path — one implementation, glyph-bundle.mjs.
const buildSet = npzBase => buildSetFromNpz(join(FONTS, npzBase));

function buildBundle() {
  const built = SETS.map(([name, npz]) => ({ name, npz, ...buildSet(npz) }));
  let dirLen = 8;
  for (const s of built) dirLen += 1 + Buffer.byteLength(s.name) + 1 + Buffer.byteLength(s.npz) + 1 + 8 + 8;
  const dir = [Buffer.from('GBF1')];
  const cnt = Buffer.alloc(4);
  cnt.writeUInt32LE(built.length, 0);
  dir.push(cnt);
  let off = dirLen;
  for (const s of built) {
    const nameB = Buffer.from(s.name), npzB = Buffer.from(s.npz);
    const e = Buffer.alloc(1 + nameB.length + 1 + npzB.length + 1 + 8 + 8);
    let p = 0;
    e.writeUInt8(nameB.length, p); nameB.copy(e, p + 1); p += 1 + nameB.length;
    e.writeUInt8(npzB.length, p); npzB.copy(e, p + 1); p += 1 + npzB.length;
    e.writeUInt8(s.linear ? 1 : 0, p); p += 1;
    e.writeDoubleLE(s.sizePx, p); p += 8;
    e.writeUInt32LE(off, p); e.writeUInt32LE(s.payload.length, p + 4);
    dir.push(e);
    off += s.payload.length;
  }
  return Buffer.concat([...dir, ...built.map(s => s.payload)]);
}

const argv = process.argv.slice(2);
const want = buildBundle();
if (argv[0] === '--check') {
  let have = null;
  try { have = readFileSync(BUNDLE); } catch {}
  if (have && have.equals(want)) {
    console.log(`ok    glyphs.bin ⇔ ${SETS.length} npz sets (${Math.round(want.length / 1024)} KB)`);
  } else {
    console.log(have ? `FAIL  glyphs.bin differs from npz rebuild (${have.length} vs ${want.length} bytes)`
      : 'FAIL  glyphs.bin missing — run: node tools/export-glyphs.mjs');
    process.exit(1);
  }
} else if (argv.length === 0) {
  writeFileSync(BUNDLE, want);
  console.log(`${BUNDLE}: ${SETS.length} sets, ${Math.round(want.length / 1024)} KB`);
} else {
  console.error('usage: node tools/export-glyphs.mjs [--check]');
  process.exit(1);
}
